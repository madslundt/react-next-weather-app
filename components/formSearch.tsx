import React, { useState, FormEvent } from "react";

const FormSearch = ({ submit, searchText }: IProps) => {
    const [term, setTerm] = useState("");

    const handleSubmit = (e: FormEvent) => {
        submit(term);
        setTerm("");

        e && e.preventDefault();
    };

    return (
        <form
            onSubmit={handleSubmit}
            method="GET"
            className="p-2 flex justify-between"
        >
            <input
                type="text"
                name="city"
                placeholder="City"
                className="border border-gray-200 rounded p-1 flex-grow"
                onChange={e => setTerm(e.currentTarget.value)}
                value={term}
                required
            />
            <button
                type="submit"
                className="hover:bg-gray-100 hover:border-gray-300 py-1 px-4 ml-1 border border-gray-200 rounded"
            >
                {searchText}
            </button>
        </form>
    );
};

interface IProps {
    submit: (term: string) => void;
    searchText?: string;
}

export default FormSearch;
