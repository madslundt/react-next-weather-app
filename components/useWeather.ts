import { useState, useEffect } from "react";

import { getWeatherByCity } from "../providers/getWeather.provider";

const useWeather = (props: IProps = {}) => {
    const [weather, setWeather] = useState<IWeather | undefined>(props.weather);
    const [isFetching, setIsFetching] = useState<boolean>(false);
    const [error, setError] = useState<IError | undefined>(props.error);
    const [cityName, setCityName] = useState<string | undefined>(
        props.cityName
    );

    const resetWeather = () => {
        setWeather(undefined);
    };

    const fetchWeather = async () => {
        try {
            if (!cityName) {
                return;
            }

            setIsFetching(true);
            setError(undefined);
            resetWeather();

            const { data } = await getWeatherByCity(cityName);

            setWeather({
                temperature: data.main.temp,
                humidity: data.main.humidity,
                windSpeed: data.wind.speed,
                windDegree: data.wind.deg
            });
        } catch (exception) {
            setError({
                message: "Unknown error",
                status: 0,
                ...(exception.response && exception.response.data)
            });
        } finally {
            setIsFetching(false);
        }
    };

    useEffect(() => {
        if (!!cityName) {
            fetchWeather();
        }
    }, [cityName]);

    return {
        weather,
        setCityName,
        cityName,
        error,
        isFetching
    };
};

export interface IProps {
    cityName?: string;
    error?: IError;
    weather?: IWeather;
}

export interface IWeather {
    temperature: number;
    humidity: number;
    windSpeed: number;
    windDegree: number;
}

export interface IError {
    message?: string;
    status: number;
}

export default useWeather;
