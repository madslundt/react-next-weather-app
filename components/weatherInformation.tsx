import React, { useMemo } from "react";
import WeatherLabel from "./weatherLabel";

const getWindDirection = (windDegree: number): WindDirection => {
    switch (windDegree) {
        case 0:
        case 360:
            return "nord";
        case 90:
            return "øst";
        case 180:
            return "syd";
        case 270:
            return "vest";
    }

    if (windDegree > 0 && windDegree < 90) {
        return "nordøst";
    } else if (windDegree > 90 && windDegree < 180) {
        return "sydøst";
    } else if (windDegree > 180 && windDegree < 270) {
        return "sydvest";
    } else if (windDegree > 270 && windDegree < 360) {
        return "nordvest";
    }

    return null;
};

const WeatherInformation = ({
    temperature,
    humidity,
    windSpeed,
    windDegree
}: IProps) => {
    const windDirection = useMemo(
        () => windDegree && getWindDirection(windDegree),
        [windDegree]
    );

    return (
        <ul>
            <li className="border-b border-panel-info p-2">
                <WeatherLabel
                    label="Temperature"
                    value={temperature && Math.round(temperature)}
                    unit="&deg;C"
                />
            </li>

            <li className="border-b border-panel-info p-2">
                <WeatherLabel
                    label="Humidity"
                    value={humidity && Math.round(humidity)}
                />
            </li>

            <li className="border-b border-panel-info p-2">
                <WeatherLabel
                    label="Wind"
                    value={windSpeed && Math.round(windSpeed)}
                    unit={`m/s ${windDirection || ""}`}
                />
            </li>
        </ul>
    );
};

interface IProps {
    temperature?: number;
    humidity?: number;
    windSpeed?: number;
    windDegree?: number;
}

export type WindDirection =
    | "nord"
    | "øst"
    | "syd"
    | "vest"
    | "nordøst"
    | "sydøst"
    | "sydvest"
    | "nordvest"
    | null;

export { getWindDirection };

export default WeatherInformation;
