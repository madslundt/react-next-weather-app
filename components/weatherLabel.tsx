import React from "react";

const WeatherLabel = ({ label, value, unit }: IProps) => (
    <span>
        {label}:
        <span className="font-bold">
            {value !== undefined && value >= 0 && (
                <>
                    &nbsp;
                    {value}
                </>
            )}
            {unit && value !== undefined && value >= 0 && (
                <>
                    &nbsp;
                    {unit}
                </>
            )}
        </span>
    </span>
);

interface IProps {
    label: string;
    value?: number;
    unit?: string;
}

export default WeatherLabel;
