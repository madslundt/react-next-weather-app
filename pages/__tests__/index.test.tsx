import React from "react";
import { shallow } from "enzyme";
import { random } from "faker";

import Home from "../index";
import { IError } from "../../components/useWeather";


describe("<Home />", () => {
    it("Should render and match snapshot", () => {
        const component = shallow(<Home cityName="cityName" />);

        component.update();

        expect(component).toMatchSnapshot();
    });

    it("Should call 'setCityName' when 'handleCityChange' is called", () => {
        // TODO
    });

    it("Should replace router with '?city={city}' when 'handleCityChange' is called", () => {
        // TODO
    });

    it("Should replace router with options 'shallow = true' when 'handleCityChange' is called", () => {
        // TODO
    });

    it("Should send props.cityName to useWeather", () => {
        // TODO
    });
    it("Should send props.weather to useWeather", () => {
        // TODO
    });
    it("Should send props.error to useWeather", () => {
        // TODO
    });
});

describe("getInitalProps", () => {
    it("Should return 'Copenhagen' as default cityName", async () => {
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {}
        });

        const actualCityName = initialProps.cityName;

        expect(actualCityName).toBe("Copenhagen");
    });

    it("Should return same cityName as query.city", async () => {
        const expectedCityName = random.word();

        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: expectedCityName
            }
        });
        const actualCityName = initialProps.cityName;

        expect(actualCityName).toBe(expectedCityName);
    });

    it("Should return 'Copenhagen' as default cityName when query.city is a list", async () => {
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: []
            }
        });
        const actualCityName = initialProps.cityName;

        expect(actualCityName).toBe("Copenhagen");
    });
    it("Should return 'Copenhagen' as default cityName when query.city is an object", async () => {
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: {} as any
            }
        });
        const actualCityName = initialProps.cityName;

        expect(actualCityName).toBe("Copenhagen");
    });
    it("Should return 'Copenhagen' as default cityName when query.city is a number", async () => {
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: random.number() as any
            }
        });
        const actualCityName = initialProps.cityName;

        expect(actualCityName).toBe("Copenhagen");
    });

    it("Should call 'getWeatherByCity' when using default cityName", async () => {
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {}
        });

        // spy
        // TODO
    });
    it("Should call 'getWeatherByCity' when using query.city", async () => {
        const expectedCityName = random.word();
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: expectedCityName
            }
        });

        // spy
        // TODO
    });
    it("Should return cityName when calling 'getWeatherByCity' successfully", async () => {
        const expectedCityName = random.word();
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: expectedCityName
            }
        });

        // spy
        // TODO
    });
    it("Should return weather when calling 'getWeatherByCity' successfully", async () => {
        const expectedCityName = random.word();
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: expectedCityName
            }
        });

        // spy
        // TODO
    });
    it("Should return cityName when calling 'getWeatherByCity' unsuccessfully", async () => {
        const expectedCityName = random.word();
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: expectedCityName
            }
        });

        // spy
        // TODO
    });
    it("Should return weather when calling 'getWeatherByCity' unsuccessfully", async () => {
        const expectedCityName = random.word();
        const initialProps = await Home.getInitialProps({
            pathname: "",
            query: {
                city: expectedCityName
            }
        });

        // spy
        // TODO
    });
});
