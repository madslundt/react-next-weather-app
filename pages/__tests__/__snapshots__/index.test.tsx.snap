// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> Should render and match snapshot 1`] = `
<div
  className="max-w-sm mx-auto my-4 bg-white overflow-hidden"
>
  <div
    className="rounded overflow-hidden border border-blue-200"
  >
    <div
      className="bg-blue-100 p-2 text-blue-300"
    >
      Weather in 
      <span
        className="font-bold"
      >
        cityName
      </span>
    </div>
    <WeatherInformation />
    <FormSearch
      searchText="Search"
      submit={[Function]}
    />
  </div>
</div>
`;
